File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    words.push(begin);
    
    let answer = words.length;
    const target_index = words.indexOf(target);
    
    if (target_index === -1) return 0;
    
    const graph = words.map(word1 => {
        const link = [];
        
        words.forEach((word2, index) => {
            let count = 0;
            
            for (let i = 0; i < word1.length; i++) {
                if (word1[i] !== word2[i]) count++;
            }
            
            if (count === 1) link.push(index);
        });
        
        return link;
    });
    
    const dfs = (node, dist, visited) => {
        if (node === target_index) {
            if (answer > dist) answer = dist;
            return;
        }
        
        visited[node] = true;
        
        graph[node].forEach(child => {
            if (!visited[child]) {
                dfs(child, dist+1, visited);
            }
        });
    };
    
    dfs(words.length-1, 0, new Array(words.length).fill(false));
    
    return answer;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    if target not in words: return 0 
    
    words.append(begin)
    
    answer = len(words)
    target_index = words.index(target)
    
    graph = []
    
    for word1 in words:
        link = []
        for index, word2 in enumerate(words):
            count = 0
            for i in range(len(word1)):
                if word1[i] != word2[i]:
                    count += 1
            if count == 1:
                link.append(index)
        graph.append(link)
    
    def dfs(node, dist, visited):
        nonlocal answer
        if node == target_index:
            if answer > dist: answer = dist
            return
        
        visited[node] = True
        
        for child in graph[node]:
            if not visited[child]:
                dfs(child, dist+1, visited)
        
    dfs(len(words)-1, 0, [False for _ in range(len(words))])
    
    return answer
